Fix cluster exclusion for egress firewall in dualstack mode

martinkennelly opened the pull request:
Previous to this change, OVN was always evaluating our cluster exclusion ACL match expression to false if we had a match expression that directly compared IP V4 and IP V6 operands. 
Example:
```
(ip4.dst == 0.0.0.0/0) && (ip4.src == $a5154718082306775057 || ip6.src == $a5154715883283518635) && ip4.dst != 10.244.0.0/16 && ip6.dst != fd00:10:244::/48" 
```
Notice at the end of that expression, we have operand "ip6.dst" AND'd with everything previous to it, therefore we always compare to this operand even if the packet is IP V4. OVN doesnt like this and will always evaluate this expression to false for IP V4.

~~This code will need a refactor in the future to consider if an exclusion for IP V6 and/or IP V4 is needed.
For example, if the match expression only contains IP V4 operands, then theres no need for IP V6 cluster exclusion operators.~~ Added

npinaeva commented:
Now we only generate subnet exclusion for cidr-based rules, which means we can save one comparison based on that cidr ip family, which in turn should save us some ovs flows, since extra `ip<X>.dst != <cluster subnet>` won't be evaluated.
From the code side: `generateMatch` pretends it can handle multiple destinations, but for CIDR rules it is always only 1, so actually we can just add cluster subnet exclusion to `func (m *matchTarget) toExpr()` like
```
func (m *matchTarget) toExpr() (string, error) {
	switch m.kind {
	case matchKindV4CIDR:
		match := fmt.Sprintf("ip4.dst == %s && %s", m.value)
		if m.clusterSubnetIntersection {
			extraMatch := getClusterSubnetsExclusion("ip4")
			match = fmt.Sprintf("%s && %s", match, extraMatch)
		}
		return match, nil
	case matchKindV6CIDR:
		match := fmt.Sprintf("ip6.dst == %s && %s", m.value)
		if m.clusterSubnetIntersection {
			extraMatch := getClusterSubnetsExclusion("ip6")
			match = fmt.Sprintf("%s && %s", match, extraMatch)
		}
		return match, nil
	case matchKindV4AddressSet:
		if m.value != "" {
			return fmt.Sprintf("ip4.dst == $%s", m.value), nil
		}
		return "", nil
	case matchKindV6AddressSet:
		if m.value != "" {
			return fmt.Sprintf("ip6.dst == $%s", m.value), nil
		}
		return "", nil
	}
	return "", fmt.Errorf("invalid MatchKind")
}
```
and remove that part from `generateMatch`
wdyt?

martinkennelly commented:
Even though OVN optimises out redundant IP protocol srcs, I also removed them in this PR.
Example: if dst is only IPv4, we only need IPv4 src constraint. 
Previous to this change, we also added in IPv6 src constraint also.

martinkennelly commented:
/retest

trozet reviewed (approved):
/lgtm

npinaeva commented:
/lgtm
